src/Dialect/ONNX/IndexExpr.hpp

```diff
@@ -458,8 +407,10 @@ class IndexExpr {
   bool isSymbol() const;
   bool isDim() const;
   bool isPredType() const;
-  bool isIndexType() const { return !isPredType(); }
-  bool isShapeInferencePass() const;
+  bool isIndexType() const;
+  bool isShapeInferencePass() const {
+    return getScope().isShapeInferencePass();
+  }
```

Comment:
When I read the fold function for canonicalization, I feel that isShapeInferencePass might be generalized to something like isForFolding. Just comment, not need to change now.

Reply:
Right, maybe the name is misleading. To indexExpr, is is more relevant to know if there is a rewriter and operations can be generated or not. Right now, inference is the only place where we cannot generate new operations, just update the type. If you have a better name, I am happy to use it.